subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/VersionControlConsumerConnection.java

```diff
@@ -0,0 +1,133 @@
+import org.gradle.tooling.internal.build.VersionOnlyBuildEnvironment;
+import org.gradle.tooling.internal.consumer.TestExecutionRequest;
+import org.gradle.tooling.internal.consumer.parameters.ConsumerOperationParameters;
+import org.gradle.tooling.internal.consumer.versioning.VersionDetails;
+import org.gradle.tooling.internal.protocol.ConnectionVersion4;
+import org.gradle.tooling.model.build.BuildEnvironment;
+import org.gradle.tooling.model.internal.Exceptions;
+import org.gradle.util.GradleVersion;
+
+import java.io.IOException;
+
+public class VersionControlConsumerConnection implements ConsumerConnection {
+    private static final String UNSUPPORTED_RUN_MESSAGE = "Support for builds using Gradle versions older than 1.2 was removed in tooling API version 3.0. You are currently using Gradle version %s. You should upgrade your Gradle build to use Gradle 1.2 or later.\n";
+    private static final String DEPRECATION_WARNING = "Support for %s older than %s is deprecated and will be removed in 5.0.%s You should upgrade your %s.\n";
+    public static final GradleVersion MIN_PROVIDER_VERSION = GradleVersion.version("1.2");
+    public static final GradleVersion MIN_LTS_PROVIDER_VERSION = GradleVersion.version("2.6");
+
+    private final ConsumerConnection delegate;
+    private final String displayName;
+    private final GradleVersion providerVersion;
+    private final boolean deprecated;
+    private final boolean unsupported;
+
+    private final ProtocolToModelAdapter adapter;
+
+    public VersionControlConsumerConnection(ConsumerConnection delegate, VersionDetails versionDetails) {
+        this(delegate, null, versionDetails.getVersion(), delegate.getDisplayName());
+    }
+
+    public VersionControlConsumerConnection(ConnectionVersion4 delegate, ProtocolToModelAdapter adapter) {
+        this(null, adapter, delegate.getMetaData().getVersion(), delegate.getMetaData().getDisplayName());
+    }
+
+    private VersionControlConsumerConnection(ConsumerConnection connection, ProtocolToModelAdapter adapter, String providerVersion, String displayName) {
+        this.delegate = connection;
+        this.providerVersion = GradleVersion.version(providerVersion);
+        this.unsupported = this.providerVersion.compareTo(MIN_PROVIDER_VERSION) < 0;
+        this.deprecated = this.providerVersion.compareTo(MIN_LTS_PROVIDER_VERSION) < 0;
+        this.displayName = displayName;
+        this.adapter = adapter;
+    }
+
+
+    @Override
+    public void stop() {
+        if (delegate != null) {
+            delegate.stop();
+        }
+    }
+
+    @Override
+    public String getDisplayName() {
+        return displayName;
+    }
+
+    @Override
+    public <T> T run(Class<T> type, ConsumerOperationParameters operationParameters) throws UnsupportedOperationException, IllegalStateException {
+        if (adapter != null && type.equals(BuildEnvironment.class)) {
+            return adapter.adapt(type, doGetBuildEnvironment());
+        }
```

Comment:
I refactor the original [`UnsupportedOlderVersionConnection`](https://github.com/gradle/gradle/blob/master/subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/UnsupportedOlderVersionConnection.java) and copy some code from there. However, this snippet seems to be doing some compatibility work - it provides a `BuildEnvironment` with only version information to something like `TestKit` (because if I remove this, some tests would fail). I doubt if this is necessary, because this piece of code seems to be very [old](https://github.com/gradle/gradle/blame/master/subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/UnsupportedOlderVersionConnection.java).